httpcore5/src/main/java/org/apache/hc/core5/http/impl/bootstrap/HttpRequester.java

```diff
@@ -261,7 +264,20 @@ private Socket createSocket(final HttpHost targetHost) throws IOException {
         }
 
         final InetSocketAddress targetAddress = addressResolver.resolve(targetHost);
-        sock.connect(targetAddress, socketConfig.getSoTimeout().toMillisIntBound());
+        // Run this under a doPrivileged to support lib users that run under a SecurityManager this allows granting connect permissions
+        // only to this library
+        try {
+            AccessController.doPrivileged(new PrivilegedExceptionAction<Object>() {
+                @Override
+                public Object run() throws IOException {
+                    sock.connect(targetAddress, socketConfig.getSoTimeout().toMillisIntBound());
+                    return null;
+                }
+            });
+        } catch (PrivilegedActionException e) {
+            // only checked exceptions are wrapped - error and RTExceptions are rethrown by doPrivileged
+            throw (IOException) e.getCause();
```

Comment:
@s1monw Even though `IOException` is the only checked exception thrown by the privileged operation, unconditional cast to  `IOException` still looks hideous. Could you please wrap it with `instance of` check and throw `IllegalStateException` in case the check fails or some such?

Reply:
can we add an assertion instead? It seems like a broken behavior and should be fatal? WDYT?

Reply:
@s1monw Sure. That would serve. Please do use `org.apache.hc.core5.util.Asserts` for consistency with the rest of the code base.

Reply:
done